client/apiversion.go

```diff
@@ -0,0 +1,183 @@
+package client
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"strings"
+
+	goosehttp "gopkg.in/goose.v1/http"
+)
+
+type apiVersion struct {
+	major int
+	minor int
+}
+
+type apiVersionInfo struct {
+	Version apiVersion       `json:"id"`
+	Links   []apiVersionLink `json:"links"`
+	Status  string           `json:"status"`
+}
+
+type apiVersionLink struct {
+	Href string `json:"href"`
+	Rel  string `json:"rel"`
+}
+
+type apiURLVersion struct {
+	rootURL          url.URL
+	serviceURLSuffix string
+	versions         []apiVersionInfo
+}
+
+// Return a full formed serviceURL based on the API version requested, the rootURL and
+// the serviceURLSuffix.  If not match to the requested API version an error is returned.
+// If only the major version is defined for the requested version, the highest minor
+// version available is assumed.
+func getAPIVersionURL(apiURLVersionInfo *apiURLVersion, requested apiVersion) (string, error) {
+	var match string
+	for _, v := range apiURLVersionInfo.versions {
+		if v.Version.major != requested.major {
+			continue
+		}
+		if requested.minor != -1 && v.Version.minor != requested.minor {
+			continue
+		}
+		for _, link := range v.Links {
+			if link.Rel != "self" {
+				continue
+			}
+			hrefURL, err := url.Parse(link.Href)
+			if err != nil {
+				return "", err
+			}
+			match = hrefURL.Path
+		}
+		if requested.minor != -1 {
+			break
+		}
+	}
+	if match == "" {
+		return "", fmt.Errorf("could not find matching URL")
+	}
+	versionURL := apiURLVersionInfo.rootURL
+	versionURL.Path = path.Join(versionURL.Path, match, apiURLVersionInfo.serviceURLSuffix)
+	return versionURL.String(), nil
+}
+
+func (v *apiVersion) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+	parsed, err := parseVersion(s)
+	if err != nil {
+		return err
+	}
+	*v = parsed
+	return nil
+}
+
+func parseVersion(s string) (apiVersion, error) {
```

Comment:
To aid the reader, a comment here saying that this function parses version strings of the form "v<major>", eg "v2", or "v<major>.<minor>, eg "v2.1"" would be great. Or words to that effect.